pkg/controller/kogitoinfra/mongodb.go

```diff
@@ -0,0 +1,290 @@
+
+	credentials := &infrastructure.MongoDBCredential{}
+	if len(i.instance.Spec.Config[infraConfigUserKey]) > 0 {
+		credentials.Username = i.instance.Spec.Config[infraConfigUserKey]
+	}
+	if len(i.instance.Spec.Config[infraConfigDatabaseKey]) > 0 {
+		credentials.Database = i.instance.Spec.Config[infraConfigDatabaseKey]
+	}
+	if len(i.instance.Spec.Config[infraConfigAuthDatabaseKey]) > 0 {
+		credentials.AuthDatabase = i.instance.Spec.Config[infraConfigAuthDatabaseKey]
+	}
+
+	err := infrastructure.FillMongoDBCredential(i.client, credentials, mongoDBInstance)
+	if err != nil {
+		return nil, err
+	}
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      mongoDBSecretName,
+			Namespace: namespace,
+		},
+		Type: corev1.SecretTypeOpaque,
+		StringData: map[string]string{
+			infrastructure.MongoDBAppSecretAuthDatabaseKey: credentials.AuthDatabase,
+			infrastructure.MongoDBAppSecretUsernameKey:     credentials.Username,
+			infrastructure.MongoDBAppSecretPasswordKey:     credentials.Password,
+			infrastructure.MongoDBAppSecretDatabaseKey:     credentials.Database,
+		},
+	}
+	if err := controllerutil.SetOwnerReference(i.instance, secret, i.scheme); err != nil {
+		return nil, err
+	}
+	if err := kubernetes.ResourceC(i.client).Create(secret); err != nil {
+		log.Errorf("Error occurs while creating %s", secret)
+		return nil, err
+	}
+	log.Debug("%s successfully created", secret)
+	return secret, nil
+}
+
+type mongoDBInfraReconciler struct {
+	targetContext
+}
+
+// getMongoDBWatchedObjects provide list of object that needs to be watched to maintain MongoDB kogitoInfra resource
+func getMongoDBWatchedObjects() []framework.WatchedObjects {
+	return []framework.WatchedObjects{
+		{
+			GroupVersion: mongodb.SchemeGroupVersion,
+			AddToScheme:  mongodb.SchemeBuilder.AddToScheme,
+			Objects:      []runtime.Object{&mongodb.MongoDB{}},
+		},
+		{
+			Objects: []runtime.Object{&corev1.Secret{}},
+		},
+	}
+}
+
+// Reconcile reconcile Kogito infra object
+func (i *mongoDBInfraReconciler) Reconcile() (requeue bool, resultErr error) {
+	var mongoDBInstance *mongodb.MongoDB
+
+	if !infrastructure.IsMongoDBAvailable(i.client) {
+		return false, errorForResourceAPINotFound(&i.instance.Spec.Resource)
+	}
+
+	// Step 1: check whether user has provided custom mongoDB instance reference
+	if len(i.instance.Spec.Resource.Name) > 0 {
+		log.Debugf("Custom mongodb instance reference is provided")
+
+		namespace := i.instance.Spec.Resource.Namespace
+		if len(namespace) == 0 {
+			namespace = i.instance.Namespace
+			log.Debugf("Namespace is not provided for custom resource, taking instance namespace(%s) as default", namespace)
+		}
+		mongoDBInstance, resultErr = i.loadDeployedMongoDBInstance(i.instance.Spec.Resource.Name, namespace)
+		if resultErr != nil {
+			return false, resultErr
+		}
+		if mongoDBInstance == nil {
+			return false, errorForResourceNotFound("MongoDB", i.instance.Spec.Resource.Name, namespace)
+		}
```

Comment:
This code is essentially the same thing as the `else`  branch. We can move it to outside of the condition.

Reply:
did not get that one ...